changefeedccl: create db-level changefeed job

KeithCh opened the pull request:
Support the creation of database-level changefeed
jobs e.g.
CREATE CHANGEFEED FOR DATABASE d;
    
We do this by modifying AllTargets to expand
database-level changefeed target spec to fetch
all tables in the database.

Resolves: #147371
Epic: CRDB-1421

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149984)
<!-- Reviewable:end -->

asg0451 reviewed (changes requested):
i think if we want to merge this, we should remove the unimplemented error so you can uncomment all those tests. that's fine by me

andyyang890 commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -37,33 +43,51 @@ type ChangefeedConfig struct {
 			if ts.DescID > 0 {
-				if ts.StatementTimeName == "" {
-					ts.StatementTimeName = cd.Tables[ts.DescID].StatementTimeName
+				if ts.Type == jobspb.ChangefeedTargetSpecification_DATABASE {
```
nit: would prefer to see a switch statement instead with a default case returning an assertion error for unhandled types

andyyang890 commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -37,33 +43,51 @@ type ChangefeedConfig struct {
+					}
+					targets, err = getTargetTargetsFromDatabaseSpec(ts, ctx, execCfg)
+					if err != nil {
+						return
```
nit: avoid naked returns for non-trivial functions

KeithCh replied:
As in I should wrap the error? Or explicitly return it?

andyyang890 replied:
Explicitly return the error along with the zero value for the other return values

andyyang890 commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -78,6 +102,47 @@ func AllTargets(cd jobspb.ChangefeedDetails) (targets changefeedbase.Targets) {
+				continue
+			}
+			// Skip system tables and virtual tables
+			if !tableDesc.IsPhysicalTable() || desc.GetParentID() == keys.SystemDatabaseID {
```
Could we do the system database check somewhere higher up? You should be able to tell that it's the system table without fetching all the tables.

KeithCh replied:
I'll move it up to `createChangefeedJobRecord`

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -622,12 +625,14 @@ func createChangefeedJobRecord(
 	} else {
 		return nil, errors.AssertionFailedf("unknown changefeed level: %s", changefeedStmt.Level)
 	}
+	tolerances := opts.GetCanHandle()
```
super nit: put this even closer to where it's used, i.e. I'd probably add a newline after the `AllTargets` call and then put it right above your for loop accumulators

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1915,7 +1935,11 @@ func logChangefeedCanceledTelemetry(ctx context.Context, job *jobs.Job) {
+	details jobspb.ChangefeedDetails,
+	description string,
+	jobID jobspb.JobID,
+	execCfg *sql.ExecutorConfig,
```
Instead of passing in the `execCfg`, could you pass the targets in?

andyyang890 replied:
A few reasons:
1. Presumably, the targets have already been computed and it'd save having to recompute them every time you need to log.
2. It reduces coupling the executor config with the logging functions given that only the targets are required; improves readability too.
3. Makes testing easier -- you don't have to create an executor config to test this function.

KeithCh replied:
I don't think it would work in this case since the number of targets can dynamically change outside of ALTER CHANGEFEED statements. If it's that costly the alternative is to not log `NumTables` at all.

andyyang890 replied:
> I don't think it would work in this case since the number of targets (what we're logging) can dynamically change.

Can they change without a changefeed restart? If so, it seems like a lot of other callsites of `AllTargets` are suspect given they don't try to fetch updated values during its lifetime (e.g. kv feed config).

KeithCh replied:
Based on the design doc it does need to restart. Ignore everything I said.

andyyang890 replied:
If it needs to restart, then we can just compute targets once then and pass it around right?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -159,6 +159,72 @@ func TestChangefeedBasics(t *testing.T) {
+		})
+	}
+
+	cdcTest(t, testFn, feedTestForceSink("kafka"))
```
nit: replace with `cdcTest(t, testFn)` before you merge this -- we don't need to run every sink all the time and explicitly specifying the sinks means any new sinks would be missed

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -897,6 +1032,14 @@ func TestMissingTableErr(t *testing.T) {
 	})
 }
 
+func TestMissingDatabaseErr(t *testing.T) {
```
Could you add another test that the system database can't be targeted?

asg0451 commented on pkg/ccl/changefeedccl/event_processing_test.go:
```diff
@@ -152,11 +152,17 @@ func TestTopicForEvent(t *testing.T) {
 				topicDescriptorCache: make(map[TopicIdentifier]TopicDescriptor),
 			}
-			tn, err := MakeTopicNamer(AllTargets(tc.details))
+			// Can pass in nil for ctx and execCfg to AllTargets because we're not testing
```
dont use a nil ctx, just use background

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -918,7 +934,7 @@ Few hours to a few days range are appropriate values for this option.`
 	}
 
-	return jr, nil
+	return jr, targets.Size, nil
```
Should we be updating `DescriptorIDs` above as well?

KeithCh replied:
oh yeah 🤦‍♂️

andyyang890 commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -37,33 +42,59 @@ type ChangefeedConfig struct {
 // version of the ChangefeedDetails protobuf.
-func makeChangefeedConfigFromJobDetails(d jobspb.ChangefeedDetails) ChangefeedConfig {
+func makeChangefeedConfigFromJobDetails(
+	d jobspb.ChangefeedDetails, ctx context.Context, execCfg *sql.ExecutorConfig,
```
ctx needs to be the first argument here too

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -247,7 +249,7 @@ func changefeedPlanHook(
 		}
 
-		jr, err := createChangefeedJobRecord(
+		jr, targetsSize, err := createChangefeedJobRecord(
```
why is it necessary to return this? cant you just do len(targets)?

KeithCh replied:
We don't have the targets in that function. We do fetch it in createChangefeedJobRecord, which is why I'm returning it from there.

KeithCh replied:
Per Andy's suggestion, I've decided to return the targets instead since it can be reused later in the function.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -498,7 +504,7 @@ func createChangefeedJobRecord(
 	jobID jobspb.JobID,
 	telemetryPath string,
-) (*jobs.Record, error) {
+) (*jobs.Record, uint, error) {
```
Hmm I don't love having to add this as a return value given it's not directly related to creating the changefeed job record.

andyyang890 replied:
Maybe you can have this function return `targets` instead and pass that around?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -75,7 +106,48 @@ func AllTargets(cd jobspb.ChangefeedDetails) (targets changefeedbase.Targets) {
+
+func getTargetsFromDatabaseSpec(
+	ts jobspb.ChangefeedTargetSpecification,
+	ctx context.Context,
```
ctx should be the first parameter here too

andyyang890 reviewed (changes requested):
Per https://github.com/cockroachdb/cockroach/pull/149984#discussion_r2216420559, if the changefeed needs to restart before targets can change, I think we should be able to just compute `AllTargets` once and pass it around.

andyyang890 commented on pkg/ccl/changefeedccl/sink.go:
```diff
@@ -249,20 +250,28 @@ func getSink(
 			return makeNullSink(&changefeedbase.SinkURL{URL: u}, metricsBuilder(nullIsAccounted))
 		case isKafkaSink(u):
 			return validateOptionsAndMakeSink(changefeedbase.KafkaValidOptions, func() (Sink, error) {
+				targets, err := AllTargets(ctx, feedCfg, serverCfg.ExecutorConfig.(*sql.ExecutorConfig))
```
You should be able to pass the targets into this function instead of needing to recompute them here (same for the other sinks in this function)

andyyang890 commented on pkg/ccl/changefeedccl/telemetry.go:
```diff
@@ -55,10 +56,11 @@ func makePeriodicTelemetryLogger(
 	description string,
 	jobID jobspb.JobID,
 	s *cluster.Settings,
+	execCfg *sql.ExecutorConfig,
```
nit: unused parameter

andyyang890 commented on pkg/ccl/changefeedccl/telemetry.go:
```diff
@@ -55,10 +56,11 @@ func makePeriodicTelemetryLogger(
 	return &periodicTelemetryLogger{
 		ctx:               ctx,
-		changefeedDetails: makeCommonChangefeedEventDetails(ctx, details, description, jobID),
+		changefeedDetails: makeCommonChangefeedEventDetails(ctx, details, description, jobID, 0),
```
Is it intentional that this is going to be zero?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -542,6 +553,11 @@ func (ca *changeAggregator) makeKVFeedCfg(
 		return kvfeed.Config{}, err
 	}
 
+	targets, err := AllTargets(ctx, ca.spec.Feed, ca.FlowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig))
```
We shouldn't need to compute `targets` twice in this function

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1924,7 +1947,11 @@ func (cf *changeFrontier) manageProtectedTimestamps(
 	// changefeed was created, it will be missing here. If so, we "migrate" it
 	// to include all the appropriate targets.
-	if targets := AllTargets(cf.spec.Feed); !makeTargetToProtect(targets).Equal(rec.Target) {
+	targets, err = AllTargets(ctx, cf.spec.Feed, cf.FlowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig))
```
IMO for both processors we should compute targets once in their `Start` functions and then store it as a field on the struct

andyyang890 reviewed (commented):
Have some more comments around minimizing the number of times we call `AllTargets`

Also I think you'll need to rebase to fix some merge conflicts

KeithCh commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -75,7 +106,46 @@ func AllTargets(cd jobspb.ChangefeedDetails) (targets changefeedbase.Targets) {
+				continue
+			}
+			var tableType jobspb.ChangefeedTargetSpecification_TargetType
+			if len(tableDesc.GetFamilies()) == 1 {
+				tableType = jobspb.ChangefeedTargetSpecification_PRIMARY_FAMILY_ONLY
+			} else {
+				tableType = jobspb.ChangefeedTargetSpecification_EACH_FAMILY
+			}
```
The table names in the changefeed would show up weird for tables that don't have families e.g. "table":"my_test_table.primary" instead of "table":"my_test_table"

KeithCh commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -75,7 +106,46 @@ func AllTargets(cd jobspb.ChangefeedDetails) (targets changefeedbase.Targets) {
+			targets.Add(changefeedbase.Target{
+				Type:              tableType,
+				DescID:            desc.GetID(),
+				StatementTimeName: changefeedbase.StatementTimeName(desc.GetName()),
```
I added a test of renaming the database watched by a db-level feed.

andyyang890 replied:
Maybe in a follow-up PR but could you also add a test where a table in the watched database is renamed?

KeithCh committed:
changefeedccl: create db-level changefeed job

Support the creation of database-level changefeed
jobs e.g.
CREATE CHANGEFEED FOR DATABASE d;

We do this by modifying AllTargets to expand
database-level changefeed target spec to fetch
all tables in the database.

Resolves: #147371
Epic: CRDB-1421

Release note: None

KeithCh commented:
bors r=andyyang890,asg0451
